src/Suave/Owin.fs

```diff
@@ -0,0 +1,621 @@
+            | Some (chCl, value) ->
+              removed := !removed |> Map.put key clock
+              true // remove after changed
+
+            | None ->
+              match initialHeaders |> Map.tryFind key with
+              | Some _ ->
+                removed := !removed |> Map.put key clock
+                true // remove from initial
+              | None ->
+                false // never present
+
+        clock <- clock + 1UL
+        res
+
+      member x.Keys = (x.Delta :> IDictionary<_, _>).Keys
+      member x.Values = (x.Delta :> IDictionary<_, _>).Values
+      member x.ContainsKey key = (x.Delta :> IDictionary<_, _>).ContainsKey key
+      member x.Add (key, value) = invalidOp "Add is not supported"
+      member x.TryGetValue (key, [<Out>] res : byref<string[]>) = (x.Delta :> IDictionary<_, _>).TryGetValue(key, ref res)
+
+    interface ICollection<KeyValuePair<string, string[]>> with
+      member x.Add kvp = invalidOp "Add is not supported"
+      member x.Count = (x.Delta :> IDictionary<_, _>).Count
+      member x.IsReadOnly = false
+      member x.Clear() = invalidOp "Clear is not supported"
+      member x.Contains kvp = (x.Delta :> ICollection<_>).Contains kvp
+      member x.CopyTo (array, arrayIndex) = (x.Delta :> ICollection<_>).CopyTo(array, arrayIndex)
+      member x.Remove kvp = (x :> IDictionary<_, _>).Remove kvp.Key
+
+    interface IEnumerable<KeyValuePair<string, string[]>> with
+      member x.GetEnumerator() = (x.Delta :> ICollection<_>).GetEnumerator()
+
+    interface IEnumerable with
+      member x.GetEnumerator() = (x.Delta :> IEnumerable).GetEnumerator()
+
+  module internal SirLensALot =
+
+    let req l =
+      HttpContext.request_ >--> l <--> untyped
+
+    let run l =
+      HttpContext.runtime_ >--> l <--> untyped
+
+    let res l =
+      HttpContext.response_ >--> l <--> untyped
+
+    let uriScheme : Property<_, string> =
+      (fun (uri : Uri) -> uri.Scheme),
+      (fun v uri -> UriBuilder(uri, Scheme = v).Uri)
+
+    let uriAbsolutePath : Property<_, _> =
+      (fun (uri : Uri) -> uri.AbsolutePath),
+      (fun v uri -> UriBuilder(uri, Path = v).Uri)
+
+    let hv2p : Property<string, string> =
+      (function | "2.0" -> "HTTP/2.0"
+                | "1.0" -> "HTTP/1.0"
+                | _ | "1.1" -> "HTTP/1.1"),
+      (fun v -> function | "HTTP/2.0" -> "2.0"
+                         | "HTTP/1.0" -> "1.0"
+                         | "HTTP/1.1" -> "1.1"
+                         | x -> x)
+
+
+    let bytesToStream : Property<byte[], IO.Stream> =
+      (fun x -> upcast new IO.MemoryStream(x)),
+      (fun v x ->
+        use ms = new IO.MemoryStream()
+        v.CopyTo ms // TO CONSIDER: save the stream instead and read from it when writing data
+        ms.ToArray())
+
+    let claimsPrincipal : Property<Map<string, obj>, ClaimsPrincipal> =
+      (fun x ->
+        x
+        |> Map.tryFind "principal" // TODO: add support for ClaimsPrincipal in Core
```

Comment:
Outstanding issue we should have a look at: supporting ClaimsPrincipal throughout core.